### PR TITLE
[ADD] sale_commission_so_based

### PR DESCRIPTION
**Pull request by maksymkv25**
An extension addon for sale_commission.

### CODE REVIEW

**Comment by maksymkv25** [comment]
@francesco-ooops Hello, I had a problem with squashing commits because in the old branch a few heads. Can you please review it PR?

**Review by francesco-ooops — APPROVED** [review]
Functional is ok!

File: sale_commission_so_based/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Copyright 2016-2022 Tecnativa - Pedro M. Baeza
```
Comment (vuwnevska) [inline sale_commission_so_based/__manifest__.py:1]: copyright from 2016 but the commit says it is ADD which means a new addon, if this addon is from another repository and just moved you need to mention this in PR and keep the original history

File: sale_commission_so_based/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+{
+    "name": "Sales commissions based on Sale Order",
+    "author": "Opsway, " "Odoo Community Association (OCA)",
```
Comment (vuwnevska) [inline sale_commission_so_based/__manifest__.py]: Tecnativa?

File: sale_commission_so_based/models/commission_settlement.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    sale_agent_line_id = fields.Many2one(comodel_name="sale.order.line.agent")
+    sale_line_id = fields.Many2one(
+        comodel_name="sale.order.line",
+        store=True,
+        related="sale_agent_line_id.object_id",
+        string="Source sale line",
+    )
+
+    @api.depends("invoice_agent_line_id", "sale_agent_line_id")
```
Comment (vuwnevska) [inline sale_commission_so_based/models/commission_settlement.py]:
```suggestion
    @api.depends("sale_agent_line_id")
```
no need to repeat the names of fields @depnds has cumulative effect

File: sale_commission_so_based/models/commission_settlement.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        # pylint: disable= missing-return
+        super()._compute_date()
+        for record in self:
+            if record.sale_agent_line_id:
+                record.date = record.sale_agent_line_id.invoice_date
+
+    @api.depends("invoice_agent_line_id", "sale_agent_line_id")
+    def _compute_commission_id(self):
+        # pylint: disable= missing-return
+        super()._compute_date()
```
Comment (vuwnevska) [inline sale_commission_so_based/models/commission_settlement.py]: maybe _compute_commission_id()?

File: sale_commission_so_based/models/commission_settlement.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        for record in self:
+            if record.sale_agent_line_id:
+                record.commission_id = record.sale_agent_line_id.commission_id
+
+    @api.depends("invoice_agent_line_id", "sale_agent_line_id")
+    def _compute_settled_amount(self):
+        # pylint: disable= missing-return
+        super()._compute_date()
+        for record in self:
+            if record.sale_agent_line_id:
```
Comment (vuwnevska) [inline sale_commission_so_based/models/commission_settlement.py:46]: also user can by mistake set both an invoice so, a good practice is to hide this possibility in UI if one option selected

File: sale_commission_so_based/models/commission_settlement.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+                record.commission_id = record.sale_agent_line_id.commission_id
+
+    @api.depends("invoice_agent_line_id", "sale_agent_line_id")
+    def _compute_settled_amount(self):
+        # pylint: disable= missing-return
+        super()._compute_date()
```
Comment (vuwnevska) [inline sale_commission_so_based/models/commission_settlement.py]:
```suggestion
        super()._compute_settled_amount()
```
?

File: sale_commission_so_based/models/sale_order.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+                raise exceptions.ValidationError(
+                    _("You can't modify a settled line"),
+                )
+
+    def _skip_settlement(self):
+        """This function should return False if the commission can be paid.
+        :return: bool
+        """
+        self.ensure_one()
+        return self.order_id.state not in ("sale", "done")
```
Comment (vuwnevska) [inline sale_commission_so_based/models/sale_order.py:69]: this part can be done configurable from settings and a list of states can be set by the user (just suggesting)

**Review by vuwnevska — COMMENTED** [review]
why was this introduced as a separate addon and not an extension to sale_commision? AFAIS it doesn't introduce conflicting functionality and the rest can be handled through settings

File: sale_commission_so_based/models/commission_settlement.py (hunk 5 above)
Comment (maksymkv25) [reply sale_commission_so_based/models/commission_settlement.py:46]: settlement line can be created only as sale_line or invoice_line, it's related to the selection field, user can't select both

File: sale_commission_so_based/models/sale_order.py (hunk 7 above)
Comment (maksymkv25) [reply sale_commission_so_based/models/sale_order.py:69]: Make sense, do it in the next iteration

File: sale_commission_so_based/models/commission_settlement.py (hunk 4 above)
Comment (maksymkv25) [reply sale_commission_so_based/models/commission_settlement.py]: of course, thanks

File: sale_commission_so_based/models/commission_settlement.py (hunk 3 above)
Comment (maksymkv25) [reply sale_commission_so_based/models/commission_settlement.py]: ok

File: sale_commission_so_based/__manifest__.py (hunk 2 above)
Comment (maksymkv25) [reply sale_commission_so_based/__manifest__.py]: Yes, added

Comment (jguenat) [reply sale_commission_so_based/__manifest__.py]: If this is a new module made by you I think there should only be you (or your company) and OCA in the author/copyright

**Review by manuelcalerosolis — APPROVED** [review]
Functional test ok!

**Pull request closed without merging**